pkg/envtest/komega/interfaces.go

```diff
@@ -0,0 +1,82 @@
+	Get(client.Object) func() error
+
+	// List returns a function that lists resources and returns the occurring error.
+	// It can be used with gomega.Eventually() like this
+	//   deployments := v1.DeploymentList{ ... }
+	//   gomega.Eventually(k.List(&deployments)).To(gomega.Succeed())
+	// By calling the returned function directly it can also be used as gomega.Expect(k.List(...)()).To(...)
+	List(client.ObjectList, ...client.ListOption) func() error
+
+	// Update returns a function that fetches a resource, applies the provided update function and then updates the resource.
+	// It can be used with gomega.Eventually() like this:
+	//   deployment := appsv1.Deployment{ ... }
+	//   gomega.Eventually(k.Update(&deployment, func (o client.Object) {
+	//     deployment.Spec.Replicas = 3
+	//     return &deployment
+	//   })).To(gomega.Scucceed())
+	// By calling the returned function directly it can also be used as gomega.Expect(k.Update(...)()).To(...)
+	Update(client.Object, UpdateFunc, ...client.UpdateOption) func() error
+
+	// UpdateStatus returns a function that fetches a resource, applies the provided update function and then updates the resource's status.
+	// It can be used with gomega.Eventually() like this:
+	//   deployment := appsv1.Deployment{ ... }
+	//   gomega.Eventually(k.Update(&deployment, func (o client.Object) {
+	//     deployment.Status.AvailableReplicas = 1
+	//     return &deployment
+	//   })).To(gomega.Scucceed())
+	// By calling the returned function directly it can also be used as gomega.Expect(k.UpdateStatus(...)()).To(...)
+	UpdateStatus(client.Object, UpdateFunc, ...client.UpdateOption) func() error
+
+	// Object returns a function that fetches a resource and returns the object.
+	// It can be used with gomega.Eventually() like this:
+	//   deployment := appsv1.Deployment{ ... }
+	//   gomega.Eventually(k.Object(&deployment)).To(HaveField("Spec.Replicas", gomega.Equal(pointer.Int32(3))))
+	// By calling the returned function directly it can also be used as gomega.Expect(k.Object(...)()).To(...)
+	Object(client.Object) func() (client.Object, error)
+
+	// ObjectList returns a function that fetches a resource and returns the object.
+	// It can be used with gomega.Eventually() like this:
+	//   deployments := appsv1.DeploymentList{ ... }
+	//   gomega.Eventually(k.ObjectList(&deployments)).To(HaveField("Items", HaveLen(1)))
+	// By calling the returned function directly it can also be used as gomega.Expect(k.ObjectList(...)()).To(...)
+	ObjectList(client.ObjectList, ...client.ListOption) func() (client.ObjectList, error)
+
+	// WithClient returns a copy that uses the given client.
+	WithClient(client.Client) Komega
+	// WithContext returns a copy that uses the given context.
+	WithContext(context.Context) Komega
+}
+
+// UpdateFunc receives an object and expects a modified version of it to be returned.
+type UpdateFunc func(client.Object) client.Object
```

Comment:
Afaik when modifying this object the "input parameter" will be modified automatically.

Is the expectation that the input client.Object is deep copied before modification?

Or maybe we don't need a return value?

I think it would be good to be clear about the expectation.

Reply:
Okay got it. So it sounds like it might be worth to see an alternative implementation with Go generics before we move forward?

Reply:
Okay saw your answer below